### DISCUSSION TITLE
When using postponed annotations (from __future__ import annotations), class based dependencies fail to resolve their dependencies with NameErrorr

### DISCUSSION THREAD

**Discussion by waweber** [fastapi/fastapi#8588] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from __future__ import annotations

from fastapi import FastAPI, Depends

app = FastAPI()

CoolType = int

def get_cool_type() -> int:
    return 123

class MyClass:
    def __init__(self, val: CoolType = Depends(get_cool_type)):
        self.val = val

def get_my_class(val: CoolType = Depends(get_cool_type)):
    return MyClass

@app.get("/")
def my_view(mycls: MyClass = Depends()):
    return mycls.val

@app.get("/test2")
def my_view_working(mycls: MyClass = Depends(get_my_class)):
    return mycls.val
```

### Description

Run the example app, and you will get a NameError because CoolType cannot be resolved. If you comment out the first view, the program will work correctly. The second view just uses a function to construct and return MyClass instead using MyClass as a dependency directly.

A possible root of the problem is in `fastapi.dependency.utils:get_typed_signature`:

```py3
def get_typed_signature(call: Callable[..., Any]) -> inspect.Signature:
    signature = inspect.signature(call)
    globalns = getattr(call, "__globals__", {})
    ...
```

To resolve the annotations, this function uses the dependency's `__globals__` property to look up the names. When `call` is a function, this works just fine. When `call` is a class, it doesn't have a `__globals__`, so we get an empty namespace as a fallback.

However, when we specify a class as a dependency, what we really care about is `MyClass.__init__`. This is a function, and it has a `__globals__` containing all the types we need to resolve its dependencies.

A possible fix for this would be to make this function check if `call` is a class and use it's `__init__` method instead.

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

Python 3.10.2

### Additional Context

Also using pydantic 1.9.0

See also this related issue: https://github.com/tiangolo/fastapi/issues/1654

**Comment by waweber** [comment]
As of version 0.78.0, this bug still exists. For anyone needing a workaround, you have two options:

1. Use a function dependency instead, e.g.
```python
def use_my_dependency(some_param: str = Query(...)):
    return MyDependency(some_param)
```
2. Have all your class-based dependencies subclass something that looks like this:
```python
class DependencyBase:
    """Class-based dependency base.

    This class serves as a workaround to https://github.com/tiangolo/fastapi/issues/4557
    """

    def __init_subclass__(cls, **kw) -> None:
        super().__init_subclass__(**kw)
        cls.__globals__ = cls.__init__.__globals__
```
This sets up the class to have the `__globals__` property that FastAPI is expecting. Your class dependencies will need to have `__init__` explicitly defined in the same scope as the dependencies they refer to.

**Comment by ndevenish** [comment]
Slightly smaller example that I think is the same issue?
```python
from __future__ import annotations

from fastapi import Request, FastAPI, Depends

app = FastAPI()

class SomeDependency:
    async def __call__(self, request: Request):
        return request

handler = SomeDependency()

@app.get("/")
def get_view(injected=Depends(handler)):
    return {"success": True}
```
Giving on 0.85.0:
```
....
  File "/path/to/site-packages/fastapi/dependencies/utils.py", line 268, in get_typed_annotation
    annotation = evaluate_forwardref(annotation, globalns, globalns)
  File "pydantic/typing.py", line 70, in pydantic.typing.evaluate_forwardref
    'ValuesView',
  File "/path/to/python/3.10.2/lib/python3.10/typing.py", line 691, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'Request' is not defined
```

There is also a third workaround: Remove `from __future__ import annotations` from the file affected by this. That of course means that you have to follow the previous rules in terms of self-referencing, not using newer union syntax before 3.10 etc.

**Comment by mastercoms** [comment]
I am experiencing this bug as well, but when using nested decorators.

**Comment by mastercoms** [comment]
From what I can tell, the original bug can be fixed by updating `globalns` in `get_typed_signature` to `vars(sys.modules[call.__module__])` when call is a class rather than a function (as mentioned in [PEP-563](https://peps.python.org/pep-0563/)).

**Comment by YuriiMotov** [comment]
Initial code example works well in current version (0.124.0) except the `/test2` endpoint that has obvious implementation mistake

```diff
def get_my_class(val: CoolType = Depends(get_cool_type)):
-    return MyClass
+    return MyClass(val=val)
```
